include_physics/ideal_MHD.h

```diff
@@ -387,14 +387,300 @@ inline void init_fields(uint ix, uint iy, uint iz, global REAL* u) {
+      cfmx = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(sq(GAMMA * um[Field_p] + BBSQm) - 4 * GAMMA * um[Field_p] * um[Field_Bx] * um[Field_Bx]))/(2 * um[Field_rho]));
+      *cl = fmin(ub[Field_ux] - cfbx, um[Field_ux] - cfmx);
+      *cr = fmax(ub[Field_ux] + cfbx, um[Field_ux] + cfmx);
+      break;
+    case 1:
+      // Y direction
+      cfby = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(sq(GAMMA * ub[Field_p] + BBSQb) - 4 * GAMMA * ub[Field_p] * ub[Field_By] * ub[Field_By]))/(2 * ub[Field_rho]));
+      cfmy = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(sq(GAMMA * um[Field_p] + BBSQm) - 4 * GAMMA * um[Field_p] * um[Field_By] * um[Field_By]))/(2 * um[Field_rho]));
+      *cl = fmin(ub[Field_uy] - cfby, um[Field_uy] - cfmy);
+      *cr = fmax(ub[Field_uy] + cfby, um[Field_uy] + cfmy);
+      break;
+    case 2:
+      // Z direction
+      cfbz = sqrt((GAMMA * ub[Field_p] + BBSQb + sqrt(sq(GAMMA * ub[Field_p] + BBSQb) - 4 * GAMMA * ub[Field_p] * ub[Field_Bz] * ub[Field_Bz]))/(2 * ub[Field_rho]));
+      cfmz = sqrt((GAMMA * um[Field_p] + BBSQm + sqrt(sq(GAMMA * um[Field_p] + BBSQm) - 4 * GAMMA * um[Field_p] * um[Field_Bz] * um[Field_Bz]))/(2 * um[Field_rho]));
+      *cl = fmin(ub[Field_uz] - cfbz, um[Field_uz] - cfmz);
+      *cr = fmax(ub[Field_uz] + cfbz, um[Field_uz] + cfmz);
+      break;
+  }// Default?
+}
+
+#else
+  #error "Define speeds to use at boundaries. Options: USE_VR_BKW / USE_VR_KUSANO."
+#endif //USE_VR_KUSANO
+
+#endif //USE_PERIODIC
 
 inline void add_surface_terms(REAL time, uint ix, uint iy, uint iz, global REAL *u, REAL *du_dt) {
 
   // For periodic boundary conditions and a single block, no surface term has to be used.
 #ifndef USE_PERIODIC
-  #error "Error in ideal_MHD.cl: Surface terms not implemented for nonperiodic boundaries!"
-#endif // USE_PERIODIC*/
+  int i;
+  REAL um[NUM_TOTAL_VARS] = {0.0};
+  get_field(ix, iy, iz, 0, 0, 0, u, um); 
+  REAL4 b_bound = b_boundary(ix, iy, iz, time);
+  REAL4 u_bound = u_boundary(ix, iy, iz, time);
+  REAL rho_bound = rho_boundary(ix, iy, iz, time); //(REAL)1;
+  REAL p_bound = p_boundary(ix, iy, iz, time);
+  REAL E_bound = compute_energy(p_bound, rho_bound, u_bound.x, u_bound.y, u_bound.z, b_bound.x, b_bound.y, b_bound.z);
+
+
+  REAL ub[NUM_TOTAL_VARS] = {0.0};
+  ub[Field_ux] = u_bound.x;
+  ub[Field_uy] = u_bound.y;
+  ub[Field_uz] = u_bound.z;
+  ub[Field_rho_ux] = u_bound.x * rho_bound;
+  ub[Field_rho_uy] = u_bound.y * rho_bound;
+  ub[Field_rho_uz] = u_bound.z * rho_bound;
+  ub[Field_rho] = rho_bound;
+  ub[Field_p] = p_bound;
+  ub[Field_E] = E_bound;
+  ub[Field_Bx] = b_bound.x;
+  ub[Field_By] = b_bound.y;
+  ub[Field_Bz] = b_bound.z;
+
+  REAL flux[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxm[NUM_CONSERVED_VARS] = {0.0};
+  REAL fluxb[NUM_CONSERVED_VARS] = {0.0};
+
+  REAL alx, arx, aly, ary, alz, arz;
+ 
+  if (check_bound_xr(ix, 1)) {
+    #ifdef USE_VR_BKW
+      calc_speed_bkw(um, ub, &alx, &arx, 0);
```

Comment:
It might be better to have only one place with such define guards and to use only one function name, e.g. `calc_hll_speeds`.